Nuclear Gradients

maxscheurer opened the pull request:
Implemented together with @Drrehn.

## Features
- [x] MP2
- [x] ADC(0)
- [x] ADC(1)
- [x] ADC(2)
- [x] ADC(2)-x
- [x] ADC(3)
- [x] CVS-ADC(1) (by @iubr)
- [x] CVS-ADC(2) (by @iubr)
- [x] CVS-ADC(2)-x (by @iubr)

## ToDo
- [x] merge on `master` (#189)
- [x] rebase on #158 
- [x] finite difference data test generation (preliminary version working)
- [x] tests:
    - [x] individual components (densities etc.)
    - [x] reference tests (preliminary version working)
- [x] rebase once #123 is in (I've added some stuff from this branch here for testing purposes, will be removed)
- [x] ~add equation numbers from papers (Rehn 2019, Levchenko 2005)~

maxscheurer committed:
getting started on gradients with psi4

fix dot product issue for bare tensors

add working gradient implementation

add gradient provider to pyscf

flake

fix excitation view tests

mfherbst commented on adcc/AmplitudeVector.py:
```diff
@@ -94,7 +94,7 @@ def __setitem__(self, index, item):
                 raise KeyError(index)
         else:
-            super().__setitem__(index, item)
+            return super().__setitem__(index, item)
```
Julia error :smile:

mfherbst commented:
Awesome :+1:

maxscheurer committed:
add basis gradient tests

lgtm and coveralls (hopefully)

more fdiff

fix

iubr committed:
started cvs-adc0 gradient

hijacked AmplitudeVector for CVS-ADC orbital response.

CVS-ADC0 working!

cvs-adc1 working

iubr committed:
added the cvs-adc2 1PDMs

added cvs-adc2 2PDMs,started orbital response (need to check if the RHS is correct)

cvs-adc2 gradient working.

cvs-adc2x working

iubr committed:
cleaned up code; fixed a few TODOs;  remaining: 1/sqrt2, antisymm, adc vs. cvs-adc rhs.

fixed 1/sqrt(2)

additional comments, RHS

switched to using symmetrise and antisymmetrise to construct density matrices

maxscheurer committed:
tweak test infrastructure

maxscheurer committed:
try to finally make damn tests work :(

maxscheurer committed:
flake 8 happy?

mfherbst commented:
Of course tests need to work and all, but I wonder if a good first aim is to just get the things already implemented merged into master and release a new version?

maxscheurer commented:
The tests are horrible to implement... ensuring highly converged eigenvectors is difficult 😞 but in principle all the implemented gradient methods are correct 👍🏻

iubr commented:
Not sure if this is a good suggestion, but maybe we could run a couple of small molecules and save the results in a text file for pytest to compare agains (this is more or less how it's done in veloxchem). I can try out the cvs-adc gradients and find an optimal/good enough setup (I guess the question is of comparing analytical vs. numerical gradients)

maxscheurer commented:
The generation of finite-difference reference data works fine already, for both generic and CVS-ADC 😊👍🏻

mfherbst commented:
> The tests are horrible to implement... ensuring highly converged eigenvectors is difficult  

Can't you lower to tolerance or introduce perturbations in the molecule / orbitals to ensure they are sufficiently distinct (to avoid rotations in invariant subspaces etc.)

maxscheurer commented:
You mean the convergence tolerance/`atol` for asserts?

You mean geometry perturbations to break symmetries? Yes, I can try that for sure 👍🏻

mfherbst commented:
> You mean the convergence tolerance/atol for asserts?

Yes. It's not amazing, but it will at least catch forgotten factors of two or switched signs.

maxscheurer committed:
all molecules hate me, just h2o for now

maxscheurer committed:
refactor gradient results, add test for correlation energy

maxscheurer commented:
@iubr, the CVS-ADC2-x don't seem to be correct right now, at least I get an error for H2O of ~1e-5 (compared to 5-point finite differences). Maybe I made some mistake while cleaning up the code... Which accuracy could you obtain in your tests? I'll try to run your original branch again at some point to get to the bottom of this.

iubr commented:
The ~1e-5 is similar to what I get with my code (in comparison to the 2-point finite difference numerical gradient). I checked it today with a 5-point finite difference and it is a little smaller, but still same order (1e-5). So if there is a bug, it would be in my module as well. 
I had discussed this difference with Patrick and we had concluded that it could be due to the tolerance/convergence threshold for the adc step (I couldn't go better than 1e-6 because of the requirement for the HF reference). Of course, this might not be the right explanation for the difference. It's, of course, possible that there is still an undiscovered bug or mistake somewhere in the code/equations. I'll make another check next week. Let's see.

maxscheurer commented:
> I had discussed this difference with Patrick and we had concluded that it could be due to the tolerance/convergence threshold for the adc step

I converge the SCF to 1e-12, and the ADC procedure is converged to 1e-10 in the energy. Given the fact that all other gradient methods so far can achieve at least 1e-7 accuracy, I'm convinced that one can achieve the same for CVS-ADC2-x . I'll try to add canonical ADC2-x gradients now and see how accurate one can get.

iubr commented:
Thanks a lot!

maxscheurer committed:
add adc2x gradients, fix cvs-adc2x energy stuff

maxscheurer committed:
adc3 amplitude response

maxscheurer committed:
correct densities

maxscheurer committed:
adc3 gradients working

mfherbst commented:
Super cool! I'll review this weekend.

How do we compare speed-wise to alternative ADC(3) gradient implementations?

iubr commented:
> 

Hej Max,

Sorry for the long silence. There have been a couple of things that I had to prioritize, so I couldn't focus on cvs-adc2x, but I have been debugging these last two weeks. I didn't find anything yet, but still have some ideas of what to try (I confirmed the excitation energy PDMs are correct and tested the lambda multipliers against numerical dipole moments -- the error is the same as in cvs-adc2). I am now re-checking all the omega multipliers, but I am not sure how much longer it will take me to get to the bottom of it.  I hope not too long. Let me see if I can do it over the weekend, otherwise please go ahead without cvs-adc2x.  What was the error in the  correlation energy TPDMs?

And thanks a lot for all the help! Super-exciting that you managed to add adc2x and adc3!!

/Iulia

maxscheurer commented:
I fixed the TPDMs for CVS-ADC(2)-x in [this](https://github.com/adc-connect/adcc/pull/124/commits/e6cba8090baa2abc0ea6105cb052a1f34612ed94#diff-d8db6a984e6ebd25e75ba9f0b0e9529403c05989ee9d45915448fc744fb88804R281), where the prefactor of `g2a.vvvv` was wrong.

The "correlation part" of the energy is tested [here](https://github.com/adc-connect/adcc/blob/c1481ee144584d270d39baf6a732212f9e9f5231/adcc/gradients/test_functionality_gradients.py#L78).

iubr commented:
Thanks! I'll look into it.

iubr commented:
> 

I double checked the equations for the TPDMs and and I do get a factor 2 in the VVVV block for cvs-adc2x. In my module, it's the only way I get the correct excitation energy. I also wrote a stand-alone code to compute the excitation energy from the density matrices and without this factor 2 I do not get the correct energy. I attach the code here as a txt file in case you'd like to have a look [cvs_adc2x_dms_from_scratch.txt](https://github.com/adc-connect/adcc/files/8055240/cvs_adc2x_dms_from_scratch.txt)
. I still have to have a more careful look at the adcc version to understand what's going on,  but next week is quite crazy for me up until Thursday. It could also be great to have a short meeting -- maybe you spot something I don't in the equations. What do you think?  

Regarding cvs-adc2x updates, over the weekend I still haven't managed to figure out the reason why analytical vs. numerical agree only up to 1e-5 and not 1e-7 (as for the other adc orders). Maybe it's best you go ahead without cvs-adc2x for the moment and we add it once this mystery is solved (which I hope will not be that much longer).

maxscheurer commented:
Thanks for the script, I'll have a look.

iubr commented:
Thanks! If you'd like an overview of the density matrices, they are from https://doi.org/10.1063/5.0058221, Appendix C, Table II, 3rd column (ADC matrix terms).

iubr commented:
Hi Max,

I just found the bug! It was a sign in one of the terms for the CV block of the lambda multipliers. See [here:](https://github.com/iubr/adcc/commit/a58ccd3886a1fded5fce96aec37e34b737ba301b). 
I tried to merge the latest version of adcc/gradients branch into my cvs_gradients branch, but for some reason I cannot get it running on my computer, so I fixed the bug using the old version of the code. If you could make the fix at your end and test cvs-adc2x again, it would be super great!! Thanks a lot!! I'll make a few more tests on my end just to make sure there's nothing else.

/Iulia

maxscheurer commented:
That's great news! I will try it out later. 👍🏻😊

iubr commented:
Thank you!! Let me know how it goes :D
(I still haven't figured out the puzzle of the factor 2 in the VVVV block between my module and adcc, but I'm planning to look into it again tomorrow)

maxscheurer commented:
I found the reason for the weird factor 2 behavior in [this](https://github.com/adc-connect/adcc/blob/c1481ee144584d270d39baf6a732212f9e9f5231/adcc/gradients/TwoParticleDensityMatrix.py#L60)
 block of code. The blocks `oooo` and `vvvv` were added _again_ in the case of CVS, such that they existed twice in the `TwoParticleDensityMatrix` object. 😁 

I found some other inconsistencies in the formation of `g2a_hf` and `g2a_oresp`, with these fixes altogether, everything works!!! 💪

mfherbst commented:
Cool! Great work.

iubr commented:
Wonderful ^-^!! Thanks a lot @maxscheurer!!

mfherbst commented on adcc/gradients/TwoParticleDensityMatrix.py:
```diff
@@ -0,0 +1,290 @@
+        with two-electron integrals.
+        ALL coefficients are already accounted for in the density matrix.
+        Two blocks are returned, the first one needs to be contracted with
+        prqs, the second one with -psqr (in Chemists' notation).
```
I don't get that comment.

mfherbst commented on adcc/gradients/__init__.py:
```diff
@@ -0,0 +1,217 @@
+@dataclass(frozen=True)
+class GradientComponents:
+    natoms: int
+    nuc: np.ndarray
```
Nuc is a bit short, maybe add a comment here to explain

maxscheurer replied:
Good idea, I think it would be good to make all attribute names a bit more verbose.

mfherbst commented on adcc/gradients/__init__.py:
```diff
@@ -0,0 +1,217 @@
+    nuc: np.ndarray
+    overlap: np.ndarray
+    hcore: np.ndarray
+    two_electron: np.ndarray
```
comment: Two electron integrals?

mfherbst commented on adcc/gradients/__init__.py:
```diff
@@ -0,0 +1,217 @@
+    def total(self):
+        """Returns the total gradient"""
+        ret = sum([self.nuc, self.overlap, self.hcore, self.two_electron])
+        if self.custom is None:
+            return ret
+        for c in self.custom:
+            ret += self.custom[c]
```
```suggestion
        if self.custom:
            return ret + sum(self.custom.values())
        else:
            return ret
```

mfherbst commented on adcc/gradients/__init__.py:
```diff
@@ -0,0 +1,217 @@
+        return self.excitation_or_mp.reference_state
+
+    @property
+    def _energy(self):
```
Why the underscore? If someone uses it it's sort of their fault, no? If you add a warning I think that's fine.

maxscheurer replied:
Good point. Will rename it to `correlation_energy` or so to make it more clear what exactly is returned.

mfherbst commented on adcc/gradients/orbital_response.py:
```diff
@@ -0,0 +1,340 @@
+    A = OrbitalResponseMatrix(hf)
+    Pinv = OrbitalResponsePinv(hf)
+    x0 = (Pinv @ rhs).evaluate()
+    lam = conjugate_gradient(A, rhs=rhs, x0=x0, Pinv=Pinv,
+                             explicit_symmetrisation=None,
+                             callback=default_print)
```
Which tolerance is used and why? Should for sure be related to the convergence used for the ADC excitation vectors ... or maybe aim for relative residual decrease?

maxscheurer replied:
I think it should rather be related to the SCF convergence criterion, because we compute the response of the HF ground state (kind of 😄). In addition, MP states don't have an additional convergence criterion, so I will make this depend on the SCF conv tol. 👍

mfherbst reviewed (commented):
Just a few comments ... obviously I have not at all checked the equations.

maxscheurer commented on adcc/backends/psi4.py:
```diff
@@ -32,6 +33,72 @@
+        g2_ao_1, g2_ao_2: relaxed two-particle density matrices
+        """
+        natoms = self.mol.natom()
+        Gradient = {}
```
Shamelessly copied from some `psi4numpy` tutorial, will change this 😄

maxscheurer commented on adcc/backends/psi4.py:
```diff
@@ -32,6 +33,72 @@
+            for p in range(3):
+                deriv2_np = np.asarray(deriv2[p])
+                Gradient["TEI"][atom, p] += np.einsum(
+                    'pqrs,prqs->', g2_ao_1, deriv2_np, optimize=True
```
True, also copy/pasted from the tutorial 😄

maxscheurer committed:
warn instead of fail

maxscheurer committed:
check for warning

maxscheurer committed:
merge master

maxscheurer committed:
switch to parametrize, restructure tests

maxscheurer committed:
fix cvs-adc2x gradients and generation

maxscheurer committed:
fix generation and tests

maxscheurer committed:
Reduce PySCF gradient contraction memory (#215)

* chore: open PR for issue 214

* implement PySCF gradient memory contraction and profiling script

* remove unused shell-batched gradient path

* extract spin sub-blocks for direct gradient AO-pair transform

Add a Tensor.export_block primitive that exports a dense rank-N
sub-block over a half-open per-axis window without materialising the
full tensor, mirroring the btod_export scatter (inverse permutation)
so it stays correct on antisymmetric same-space blocks.

Rewrite to_ao_pair_density to extract only the spin sub-block each
spin case needs via export_block and contract it with the compacted
non-zero-row coefficients, instead of densifying full zero-padded TPDM
blocks and full rank-4 AO tensors. Spin sub-blocks are pre-extracted
once and reused across pair chunks; the packed density is written once
per chunk to keep the out-of-core path I/O-efficient.

Matches the full-AO path to machine precision (RHF/UHF/CVS, cc-pVDZ/
pVTZ, memory and hdf5 storage). Peak RSS at cc-pVTZ drops ~25 percent
(in-memory) to ~37 percent (out-of-core), with the win growing with
basis size. Adds export_block and spin-subblock regression tests.

* address review findings: bound hdf5 memory, drop outcore alias, remove dead code

- hdf5 storage now picks a bounded default pair_chunk_size so the
  in-memory working buffer no longer grows to the full (nao, nao, npair)
- remove the undocumented 'outcore' alias; validate storage against an
  explicit {memory, hdf5} set with a matching error message
- guard the hdf5 cleanup so it cannot mask an in-flight exception
- drop dead code: numpy export_block fallback, redundant provider
  ao_pair_density_from_dense wrapper, unused _empty_gradient keys,
  unused cblk_size in Tensor::export_block
- fold duplicated direct/exchange spin-case loops; drop redundant copy;
  hoist loop-invariant np.any(offdiag)
- fix flake8 blank-line/E402 violations
- add tests for hdf5-vs-memory equivalence, scratch cleanup, bounded
  default chunk, and the new validation guards

* reframe TPDM AO-pair packing docstrings as backend-neutral

Describe the lower-triangular symmetric AO-pair packing as a generic
convention rather than a PySCF assumption; PySCF aosym="s2kl" is now
mentioned only as one concrete consumer example.

* rename gradient TEI->ERI and add nuclear gradients docs

Rename the two-electron-integral terminology to ERI throughout the
nuclear gradient API to match adcc's established convention: the
eri_contraction / eri_shell_chunk_size / eri_pair_chunk_size /
eri_pair_density_storage kwargs of nuclear_gradient, the PySCF
_contract_eri_with_packed_density method and Gradient["ERI"] key, and
the Psi4 Gradient["ERI"] key. Tests and the profiling script follow.

Add a Nuclear gradients documentation page (usage-first, followed by a
theory and algorithm section), a supported-methods table, and citations
to Rehn and Dreuw (2019) and Brumboiu et al. (2021).

* bump test data to 0.7.4 to fix download

The 0.7.2 test-data path on the upstream server is no longer available,
causing the testdata download in conftest.py to fail at pytest collection
time (and thus all CI test jobs). Pull update_testdata.sh (SOURCE bump to
0.7.4) and the regenerated SHA256SUMS from master.

* fix: add missing return in GradientComponents.total and expand gradient contraction tests

maxscheurer committed:
Integrate master refactor into gradients branch (#224)

* chore: open PR for issue 219

* Merge origin/master into gradients branch

Integrate the master refactor (StateView/ElectronicStates restructure,
libadcc/ -> libadcc_src/ rename, pyproject.toml migration) into the
gradients feature branch.

Key changes for gradient code compatibility:
- Add forwarding properties (ground_state, reference_state, method) on
  Excitation for gradient code access
- Add total_energy property to ElectronicStates and StateView using the
  modern MethodLevel enum pattern
- Use OneParticleDensity (not OneParticleOperator) for gradient density
  matrices — PR 200 separated these with different AO transforms
  (densities: C^T D C, operators: S C^T D C S)
- Use OperatorSymmetry.NOSYMMETRY for CVS density matrices that require
  non-canonical block assignments (co/oc), with explicit transpose blocks
- Update excitation_property_keys blacklist to exclude forwarding props
- Modernize method.level usage to method.level.to_int() in tests
- Regenerate libadcc.pyi with export_block signature

Closes #219

* add backend-free total_energy and forwarding-property tests

Addresses review findings 2 and 3 on PR 224:
- test_total_energy (adc0/adc2) independently validates the total_energy
  formula across both MethodLevel branches and the Excitation/StateView
  delegation, instead of comparing against a self-generated snapshot.
- test_excitation_forwarding_properties asserts ground_state,
  reference_state and method forward to the parent ElectronicStates.

* fix CI: restore master test-data source and clear flake8 errors

The master merge accidentally reverted the test-data sync to the old
gradients-branch versions:
- update_testdata.sh SOURCE reverted from 0.7.5 back to 0.7.4
- SHA256SUMS *_mpdata.hdf5 hashes reverted to 0.7.4 values

The 0.7.4 server no longer hosts ch2nh2_sto3g_hfdata.hdf5, so pytest
collection failed with 'missing test data file'. Restore both files to
match origin/master.

Also clear pre-existing flake8 errors blocking the Code style job:
unused imports (unittest, itertools) and E302 in
functionality_gradients_test.py, and E501/W293 in dump_fdiff_gradient.py.

* fix CI test failures from master merge

Three merge regressions where gradient-branch code/tests conflicted with
master's state:

- LazyMp_test: master's mp2_diffdm produces the canonical 'oc' block
  (o1o2), and the 0.7.5 reference data stores dm_o1o2 for the adcc
  source (dm_o2o1 for adcman). Restore master's block list and the
  adcman-specific label remapping (fixes KeyError 'dm_o2o1').

- workflow: revert the gradient-branch 'warn instead of fail' changes
  back to master's strict InputError behaviour for too-low conv_tol and
  too-few guesses (workflow.py and workflow_test.py now match master).

- functionality_gradients_test: master's strict guess validation rejects
  requesting more states than available. h2o/sto3g cvs-adc0/adc1 only
  support 2 CVS singlets, so cap n_singlets to the number of reference
  states (len(energy_ref)) instead of the hard-coded 5.

maxscheurer committed:
Speed up restricted AO-pair TPDM transform (#225)

* chore: open PR for issue 220

* implement restricted pair-density fast path

* strengthen RHF fast-path tests from review findings 2 and 3

maxscheurer committed:
Record merge of master into gradients

maxscheurer committed:
Add geometry optimization scanner for gradients (#226)

* chore: open PR for issue 221

* implement geometry optimization scanner core

* refactor scanner API to accept PySCF SCF object

* fix scanner root tracking and add MP2+ADC2 excited example

Copy the PySCF Mole snapshot in the tracking descriptor so previous-step
AO bases are not silently mutated by the SCF scanner; add a regression
test for descriptor mol independence.

Rework the excited-state example to optimise on the MP2 ground-state
surface first and then run the ADC(2) excited-state optimisation from
the relaxed geometry using cc-pVDZ.

* switch excited example to s-trans-butadiene/6-31G*

Use the s-trans-butadiene geometry optimisation showcase from Rehn &
Dreuw (J. Chem. Phys. 150, 174110, 2019) with the 6-31G* basis.

* add scanner unit and geomeTRIC end-to-end gradient scanner tests

Extend the scanner unit tests with unconverged-SCF handling, an
excited-state match against the explicit run_adc + nuclear_gradient
loop, fixed-index and density-overlap root tracking, tracking-loss and
ambiguity guards, and SCF guess continuity.

Add functionality_geomopt_test.py with an end-to-end geomeTRIC ground-
state optimisation reaching a stationary point (gradient ~ 0) and a
calc_new custom-engine contract check; both skip cleanly when PySCF or
geomeTRIC is unavailable.

* harden scanner: reject MP3 gradients, observable root tracking, eager validation

Address review findings on PR 226:
- Reject ground-state MP targets with level != 2 (no MP3 gradient exists)
- Make root-tracking diagnostics observable on last_tracking every call
- Narrow _safe_ao_ndarray to expected unavailable-operator exceptions
- Validate follow eagerly at construction
- Warn on ignored run_adc kwargs for the string-MP target spelling
- Add root-reorder, excited-state e2e, and validation/scoring tests

* fix flake8: silence intended scanner re-export F401, wrap long example line

* declare gradients __all__ instead of noqa for scanner re-exports
